### PR TITLE
names(.SD) should work

### LINKED ISSUE #795: names(.SD) := ... should work
This would allow the following type of code:
`DT[, names(.SD) := lapply(.SD, rev), .SDcols = -c(1,3,8)]`
To reverse every column except 1, 3, and 8 in `DT` by reference.  See discussion in #787.  Maybe potentially even:
`DT[, := lapply(.SD, rev), .SDcols = -c(1,3,8)]`
and have the names of the columns to be updated inferred from the names of the return value to `lapply`.

### PR DESCRIPTION
**Pull request by ColeMiller1**
Closes #795. Towards #3189. 

This is the proposed implementation:
```
dt[, names(.SD) := lapply(.SD, '*', 5)]
```
- [x] Update Reference Semantics Vignette
- [x] Update .SD Usage Vignette
- [x] Update ```assign``` help
- [x] Include tests

TO DO after release:
Update [StackOverflow post](https://stackoverflow.com/questions/8508482/what-does-sd-stand-for-in-data-table-in-r/47406952#47406952) to mirror updated .SD vignette.

### CODE REVIEW

File: R/data.table.R [hunk 1]
```diff
@@ -1027,10 +1027,14 @@ replace_dot_alias = function(e) {
           lhs = jsub[[2L]]
           jsub = jsub[[3L]]
           if (is.name(lhs)) {
-            lhs = as.character(lhs)
+            if (deparse(lhs) == '.SD') lhs = sdvars else lhs = as.character(lhs)
```
Comment (jangorecki) [inline R/data.table.R]: checking against `as.name(".SD")` should be preferred rather than `deparse` which has more corner cases to handle

File: inst/tests/tests.Rraw [hunk 2]
```diff
@@ -16740,3 +16740,10 @@ test(2131, lapply(x[ , list(dt = list(.SD)), by = a]$dt, attr, '.data.table.lock
 ########################
 #  Add new tests here  #
 ########################
+
+## names(.SD) - issue #795
+DT <- data.table(a=1:6, b=1:6, c=rep(c(T,F), 3))
+mycols <- 1:2
+test(2131.1, DT[, .SD := lapply(.SD, `*`, 2), .SDcols = mycols], data.table(a = (1:6)*2, b = (1:6)*2, c = rep(c(T, F), 3)))
```
Comment (jangorecki) [inline inst/tests/tests.Rraw]: T/F into TRUE/FALSE

File: inst/tests/tests.Rraw [hunk 3]
```diff
@@ -16740,3 +16740,10 @@ test(2131, lapply(x[ , list(dt = list(.SD)), by = a]$dt, attr, '.data.table.lock
...
+test(2131.3, DT[, .SD := lapply(.SD, '*', 2), .SDcols = -3L], data.table(a = (1:6)*4, b = (1:6)*4, c = rep(c(T, F), 3)))
+test(2131.5, DT[, .SD := lapply(.SD, '*', 2)], data.table(a = (1:6)*8, b = (1:6)*8, c = rep(c(T, F), 3) * 2))
```
Comment (jangorecki) [inline inst/tests/tests.Rraw]: Multiply logical vector is quite an unfortunate use case because, unless multiplied by 0, it won't change the value. Much better to test `mycost = 2:3` and then we can ensure that first column is not being changed.

File: man/assign.Rd [hunk 4]
```diff
@@ -18,6 +18,9 @@
 #            LHS2 = RHS2,
 #            ...), by = ...]
 
+# 3. Multiple columns in place
+# DT[i, .SD = lapply(.SD, fx), by = ..., .SDcols = ...]
```
Comment (jangorecki) [inline man/assign.Rd]: there are not unit test for using `.SD` lhs combined with `i`, and similarly with `by`.

**Comment by ColeMiller1** [comment]
Thank you, Jan. I am working on better test cases that do not involve logicals. I will update tonight.

I also noticed that #4031 would be partially addressed:
```
dt = data.table(iris)
keep = c('Species', 'Sepal.Width')
dt[, .SD := NULL, .SDcols = !keep]
```
I will try to substitute ```names(.SD)``` with the ```sdvars``` - that would allow solutions like yours ```setdiff(names(.SD), keep)``` work and would provide the option of ```paste0(names(.SD), 'max')``` so you can update all of the columns without having to use ```.SDcols``` or an intermediate character vector.

File: inst/tests/tests.Rraw (hunk 3 above)
Comment (ColeMiller1) [reply inst/tests/tests.Rraw]: Multiplying a logical vector by a number changes the value on my computer: 
```
data.table(a = c(TRUE, FALSE))[, a := 2 * a][]
       a
   <num>
1:     2
2:     0
```
I agree that the tests needed improved and made some changes.

File: R/data.table.R [hunk 5]
```diff
@@ -1027,14 +1027,14 @@ replace_dot_alias = function(e) {
...
-            # if (deparse(lhs) == 'names(.SD)') {
-              # lhs = sdvars
-            # } else {
-              # e.g. (MyVar):= or get("MyVar"):=
-              lhs = eval(lhs, parent.frame(), parent.frame())
-            # }
+            #i.e lhs is names(.SD) || setdiff(names(.SD), cols) || (cols)
+            if (lhs[[1]] == as.name('names') && lhs[[2]] == as.name('.SD')) lhs = sdvars
+            for (i in seq_along(lhs)[-1]){
+              if (lhs[[i]] == as.name('names(.SD)')) lhs[[i]] = sdvars
```
Comment (jangorecki) [inline R/data.table.R]: `as.name('names(.SD)'))` AFAIK this is invalid, `as.name` is appropriate for a scalar character name of a variable or function. Here you already do `names` call, so `as.call(list(as.name("names"), as.name(".SD")))`

Comment (ColeMiller1) [reply R/data.table.R]: I was surprised that ```as.name('names(.SD)')``` worked, mainly because ```is.name(quote(names(.SD)))``` is *FALSE* but somehow ```quote(names(.SD)) == as.name('names(.SD)')``` is *TRUE*.

I looked into ```quote(names(.SD))``` as well. Using both the ```quote(...)``` method and the ```as.call(list(...))``` methods result in warnings on tests. Using ```as.name(...)``` resulted in no errors on my home PC.
 
Here's an example of one warning:
```
Running test id 352.1      Test 352.1 produced 1 warnings but expected 0
Expected: 
Observed: longer object length is not a multiple of shorter object length
###test 352.1 for reference
test(352.1, DT[,letters[1:4]:=list(1L,NULL)], error="Supplied 4 columns to be assigned 2 items. Please see NEWS for v1.12.2")
```
I also just worked out a recursive function similar to ```replace_dot_alias``` that does not produce errors or warnings and seems more appropriate. That would turn the PR into:
```
if (is.name(lhs)) {
  if (lhs == as.name('.SD')) lhs = sdvars else lhs = as.character(lhs)
} else {
  replace_names_sd = function(e){
    if (length(e) == 1L) return(e)
    if(e[[1L]] == as.name('names') && e[[2L]] == as.name('.SD')) return(sdvars)
    for (i in seq_along(e)[-1L]) if (!is.null(e[[i]])) e[[i]] = replace_names_sd(e[[i]])
    e
  }
  lhs = eval(replace_names_sd(lhs), parent.frame(), parent.frame())
}
```

Comment (jangorecki) [reply R/data.table.R]: AFAIU using `as.name('names(.SD)')` will be more or less like doing `data.table(`as.name('names(.SD)')` = 1:10)`, this has to be call rather than symbol, which is a product of `as.name`.

Comment (ColeMiller1) [reply R/data.table.R]: I made additional changes that should be more robust by using a recursive function to check all calls within LHS for ```expr[[1L]] == as.name('names') && expr[[2L]]== as.name('SD')```.

**Comment by ColeMiller1** [comment]
This should be good to go unless we want to support ```ncol(.SD)``` or ```length(.SD)``` which may be expected to work on the LHS of ```:=```. I could see that being useful for ```lapply``` with multiple functions.

Regarding codecov, I am not sure what to do. There are 10 new tests to address 7 lines of new code plus all previous tests using ```lhs := rhs``` would also use this code.

File: man/assign.Rd (hunk 4 above)
Comment (jangorecki) [inline man/assign.Rd]: There is typo, you wanted `.SD := lapply`, not `.SD = lapply`. I actually prefer to support only `names(.SD)`, but will comment on that separately.

**Comment by ColeMiller1** [comment]
Should be good again. The .SD vignette could be updated with a few items such as ```patterns()``` and the ability of using functions. I am happy to update the .SD vignette with this PR, but am unsure since they were done outside of this PR. 

But there are two examples that this and Michael's recent PR will work great with:

```
fkt_idx = which(sapply(Teams, is.factor))
Teams[ , (fkt_idx) := lapply(.SD, as.character), .SDcols = fkt_idx]
## to
 Teams[, names(.SD) := lapply(.SD, as.character), .SDcols = is.factor]

##From 
team_idx = grep('team', names(Teams), value = TRUE)
Teams[ , (team_idx) := lapply(.SD, factor), .SDcols = team_idx]
## To
Teams[, names(.SD) := lapply(.SD, factor), .SDcols = patterns('team')]

File: vignettes/datatable-sd-usage.Rmd [hunk 6]
```diff
@@ -73,7 +73,13 @@ The first way to impact what `.SD` is is to limit the _columns_ contained in `.S
 Pitching[ , .SD, .SDcols = c('W', 'L', 'G')]
 ```
 
-This is just for illustration and was pretty boring. But even this simply usage lends itself to a wide variety of highly beneficial / ubiquitous data manipulation operations:
+This is just for illustration and was pretty boring. In addition to accepting a character vector, `.SDcols` also accepts:
+
+1. a function such as `is.character`
+2. name searching with `patterns()`
+3. Integer and boolean vectors.
```
Comment (jangorecki) [inline vignettes/datatable-sd-usage.Rmd]: Integer could be lowercase, and instead of boolean it is better to use "logical", and we don't need dot at the end

File: vignettes/datatable-sd-usage.Rmd [hunk 7]
```diff
@@ -87,52 +93,40 @@ We notice that the following columns are stored as `character` in the `Teams` da
...
 For example, we could do the following to convert all `factor` columns to `character`:
 
 ```{r sd_as_logical}
-# while .SDcols accepts a logical vector,
-#   := does not, so we need to convert to column
-#   positions with which()
-fkt_idx = which(sapply(Teams, is.factor))
-Teams[ , (fkt_idx) := lapply(.SD, as.character), .SDcols = fkt_idx]
-head(unique(Teams[[fkt_idx[1L]]]))
+fct_idx = Teams[, which(sapply(.SD, is.factor))] #column numbers to show the class changing
```
Comment (jangorecki) [inline vignettes/datatable-sd-usage.Rmd]: `#column` -> `# column`

File: man/assign.Rd [hunk 8]
```diff
@@ -18,6 +18,9 @@
 #            LHS2 = RHS2,
 #            ...), by = ...]
 
+# 3. Multiple columns in place
+# DT[i, names(.SD) := lapply(.SD, fx), by = ...]
```
Comment (MichaelChirico) [inline man/assign.Rd]: example should use `.SDcols`, right? or at least mention it

File: vignettes/datatable-reference-semantics.Rmd [hunk 9]
```diff
@@ -258,14 +258,30 @@ head(flights)
...
 # RHS gets automatically recycled to length of LHS
 flights[, c("speed", "max_speed", "max_dep_delay", "max_arr_delay") := NULL]
 head(flights)
 ```
 
+#### -- How can we update multiple existing columns in place using `.SD`?
+
+```{r}
+char_cols <- sapply(flights, is.character)
+flights[, names(.SD) := lapply(.SD, as.factor), .SDcols = char_cols]
```
Comment (MichaelChirico) [inline vignettes/datatable-reference-semantics.Rmd]: Could also be a good place to show off `.SDcols = function` like:

```
flights[ , names(.SD) := lapply(.SD, as.factor), .SDcols = is.character]
```

File: R/data.table.R [hunk 10]
```diff
@@ -1029,8 +1029,14 @@ replace_dot_alias = function(e) {
           if (is.name(lhs)) {
             lhs = as.character(lhs)
           } else {
-            # e.g. (MyVar):= or get("MyVar"):=
-            lhs = eval(lhs, parent.frame(), parent.frame())
+            #i.e lhs is names(.SD) || setdiff(names(.SD), cols) || (cols)
+            replace_names_sd = function(e, cols){
```
Comment (MichaelChirico) [inline R/data.table.R]: not essential to this PR, but we have a couple of these little helper functions to iterate over the AST and edit it somehow, perhaps we can write something like `ast_apply` that can be used more generally?

File: vignettes/datatable-sd-usage.Rmd [hunk 11]
```diff
@@ -73,7 +73,13 @@ The first way to impact what `.SD` is is to limit the _columns_ contained in `.S
 Pitching[ , .SD, .SDcols = c('W', 'L', 'G')]
 ```
 
-This is just for illustration and was pretty boring. But even this simply usage lends itself to a wide variety of highly beneficial / ubiquitous data manipulation operations:
+This is just for illustration and was pretty boring. In addition to accepting a character vector, `.SDcols` also accepts:
+
+1. any function such as `is.character` to filter _columns_
+2. the function^{*} `patterns()` to filter by _column names_
```
Comment (MichaelChirico) [inline vignettes/datatable-sd-usage.Rmd]: note to self to check the GL CI output to make sure the `^{*}` looks as intended on output

**Comment by MichaelChirico** [comment]
Looks good overall, @ColeMiller1 are we ready to remove WIP label?

File: R/data.table.R [hunk 12]
```diff
@@ -1029,8 +1029,14 @@ replace_dot_alias = function(e) {
...
+              if (length(e) == 1L) return(e)
+              if (e[[1L]] == quote(names) && e[[2L]] == quote(.SD)) return(cols)
```
Comment (MichaelChirico) [inline R/data.table.R]: I think we need an `is.name(e[[2L]])` or this would apply to `names(.SD(4))`?

File: man/assign.Rd (hunk 8 above)
Comment (jangorecki) [reply man/assign.Rd]: unless it means all columns, but better to provide the argument to make example more complete

File: R/data.table.R (hunk 12 above)
Comment (ColeMiller1) [reply R/data.table.R]: Added although may not be necessary. Regardless, should be more robust.
```
quote(names(.SD(4)))[[2L]]
## .SD(4)
quote(names(.SD(4)))[[2L]] == quote(.SD)
## [1] FALSE
```

File: vignettes/datatable-reference-semantics.Rmd (hunk 9 above)
Comment (ColeMiller1) [reply vignettes/datatable-reference-semantics.Rmd]: I left it in but then in the next example, I changed it to ```.SDcols = is.factor```. Sometimes a person may want to keep track of their original column classes so it could be good to show people that they have the option. 

Still, I have no heartburn if you want the first part changed to ```.SDcols = is.character``` - it would be less information overload to just do it one way throughout the example.

File: R/data.table.R (hunk 10 above)
Comment (ColeMiller1) [reply R/data.table.R]: I am not sure what AST is. I know @jangorecki has taken ```replace_dot_alias``` to C in his #4174 PR but it is not generalized.

**Comment by ColeMiller1** [comment]
@MichaelChirico Do you know why this PR and #4212 have had trouble with the code coverage? Otherwise, I think it's good to remove WIP.

File: vignettes/datatable-sd-usage.Rmd [hunk 13]
```diff
@@ -73,7 +73,15 @@ The first way to impact what `.SD` is is to limit the _columns_ contained in `.S
 Pitching[ , .SD, .SDcols = c('W', 'L', 'G')]
```
Comment (ColeMiller1) [inline vignettes/datatable-sd-usage.Rmd]: Update the Stack Overflow post ???

https://stackoverflow.com/questions/8508482/what-does-sd-stand-for-in-data-table-in-r/47406952#47406952

File: R/data.table.R (hunk 10 above)
Comment (MichaelChirico) [reply R/data.table.R]: AST=abstract syntax tree (I actually had to look that up myself)

https://adv-r.hadley.nz/expressions.html

not something that needs to be handled in this PR, just something I wanted to write down for safe keeping

**Comment by MichaelChirico** [comment]
I don't see any coddcov issue (I see 4 checks passed)

maybe because the comment from the codecov bot didn't update/repost?

**Comment by ColeMiller1** [comment]
It says 
>No report found to compare against

And further up it says 

>!No coverage uploaded for pull request base (master@8e1c109). Click here to learn what that means.
The diff coverage is 100%.

**Comment by MichaelChirico** [comment]
I see, that wasn't being surfaced on mobile.

Not sure what's happening there, it'll probably go away on a re-trigger of the jobs. My best guess is it's been > 2 weeks since the last commit to `master`, and the diff has staled?

**Comment by ColeMiller1** [comment]
@MichaelChirico I merged master and the codecov is good now. The WIP can be removed if you still agree.

**Comment by pkress** [comment]
Any idea when this might get implemented/added to CRAN?

**Comment by MichaelChirico** [comment]
@ColeMiller1 would you like to update this PR for review targeting the subsequent release?

**Comment by ColeMiller1** [comment]
@MichaelChirico sure thing - I will work on it next weekend.

**Comment by MichaelChirico** [comment]
> I know you are a master with Git commands

Oh, far, far from it. The issue is that GitHub's conflict resolution UI is really dinky and hasn't been updated in like a decade. In such cases we have to fix the merge conflicts ourselves. VSCode has a much nicer UI for this, I guess GitHub will eventually integrate that better to the PR flow.

Anyway, the conflicts didn't look so bad to resolve, so I went ahead and tried. PTAL if anything is out of whack.

File: R/data.table.R [hunk 14]
```diff
@@ -1122,8 +1122,14 @@ replace_dot_alias = function(e) {
           if (is.name(lhs)) {
             lhs = as.character(lhs)
           } else {
-            # e.g. (MyVar):= or get("MyVar"):=
-            lhs = eval(lhs, parent.frame(), parent.frame())
+            #i.e lhs is names(.SD) || setdiff(names(.SD), cols) || (cols)
```
Comment (MichaelChirico) [inline R/data.table.R:1125]:
```suggestion
            # i.e lhs is names(.SD) || setdiff(names(.SD), cols) || (cols)
```

File: R/data.table.R [hunk 15]
```diff
@@ -1122,8 +1122,14 @@ replace_dot_alias = function(e) {
...
+            replace_names_sd = function(e, cols){
+              if (length(e) == 1L) return(e)
+              if (e[[1L]] == quote(names) && is.name(e[[2L]]) && e[[2L]] == quote(.SD)) return(cols)
```
Comment (MichaelChirico) [inline R/data.table.R:1128]: I think this will cause trouble if it's a namespaced call in LHS. We don't need to support it, but should be aware of the possibility and test:

```r
DT[, base::names(.SD) := ...]
```

```suggestion
              if (e %iscall% "names" && is.name(e2 <- e[[2L]]) && e2 == ".SD") return(cols)
```

File: R/data.table.R [hunk 16]
```diff
@@ -1122,8 +1122,14 @@ replace_dot_alias = function(e) {
...
+              for (i in seq_along(e)[-1L]) if (!is.null(e[[i]])) e[[i]] = replace_names_sd(e[[i]], cols)
```
Comment (MichaelChirico) [inline R/data.table.R:1129]:
```suggestion
              for (i in 2:length(e)) if (!is.null(e[[i]])) e[[i]] = replace_names_sd(e[[i]], cols)
```

File: inst/tests/tests.Rraw [hunk 17]
```diff
@@ -16704,6 +16704,9 @@ dt = data.table(x = rep(1:3, each = 3), y = runif(9))
 out = dt[, list(evaluated = list(f(copy(.SD)))), by = x]
 test(2131.2, class(out$evaluated[[1L]]), 'environment')
 
+
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:16707]: Looks like a bad merge, remove these lines

File: inst/tests/tests.Rraw [hunk 18]
```diff
@@ -18294,3 +18297,27 @@ test(2246.1, DT[, data.table::shift(b), by=a], DT[, shift(b), by=a], output="GFo
...
+dt = data.table(a = 1:4, b = 5:8, grp = c('a', 'a', 'b', 'c'))
+test(2247.7, dt[, names(.SD) := lapply(.SD, max), by = grp], data.table(a = c(2L, 2L, 3L, 4L), b = c(6L, 6L, 7L, 8L), grp = c('a', 'a', 'b', 'c')))
+
+dt = data.table(a = 1:4, b = 5:8, grp = c('a', 'a', 'b', 'c'))
+keep = c('a', 'b')
+test(2247.8, dt[, names(.SD) := NULL, .SDcols = !keep], data.table(a = 1:4, b = 5:8))
+
+dt = data.table(a = 1:4, b = 5:8, grp = c('a', 'a', 'b', 'c'))
+test(2247.9, dt[, paste(names(.SD), 'max', sep = '_') := lapply(.SD, max), by = grp] , data.table(a = 1:4, b = 5:8, grp = c('a', 'a', 'b', 'c'))[, paste(c('a','b'), 'max', sep = '_') := lapply(.SD, max), by = grp])
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:18319]:
```suggestion
test(2247.9, dt[, paste(names(.SD), 'max', sep = '_') := lapply(.SD, max), by = grp] , data.table(a = 1:4, b = 5:8, grp = c('a', 'a', 'b', 'c'), a_max = c(2L, 2L, 3L, 4L), b_max = c(6L, 6L, 7L, 8L)))
```

File: inst/tests/tests.Rraw [hunk 19]
```diff
@@ -18294,3 +18297,27 @@ test(2246.1, DT[, data.table::shift(b), by=a], DT[, shift(b), by=a], output="GFo
...
+
+dt = data.table(a = 1:3, b = 5:7, grp = c('a', 'a', 'b'))
+test(2247.91, dt[1:2, paste(names(.SD), 'max', sep = '_') := lapply(.SD, max), by = grp], data.table(a = 1:3, b = 5:7, grp = c('a', 'a', 'b'), a_max = c(2L, 2L, NA_integer_), b_max = c(6L, 6L, NA_integer_)))
+test(2247.92, dt[, names(.SD(2)) := lapply(.SD, .I)], error = "could not find function \".SD\"")
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:18323]: I can sort of see why `.91` is closely related to `.9`, but not so for `.92`. Let's just keep a strict ordering: `2247.01`, ..., `2247.09`, `2247.10`, `2247.11`.

Comment (MichaelChirico) [inline inst/tests/tests.Rraw:18323]:
```suggestion
test(2247.92, dt[, names(.SD(2)) := lapply(.SD, .I)], error = 'could not find function ".SD"')
```

**Review by MichaelChirico — COMMENTED** [review]
IINM we're missing one more test, namely a case where `.SD` is used _as well as some columns not in `.SD`, e.g.

```r
DT[, names(.SD) := lapply(.SD, \(x) x / b), .SDcols = "a"]
```

That way we make sure we're picking up the right ".SD" since in that case the "actual" .SD gets augmented with `b` to make sure it's around during eval.

File: vignettes/datatable-reference-semantics.Rmd [hunk 20]
```diff
@@ -250,14 +250,30 @@ head(flights)
 * The `LHS := RHS` form allows us to operate on multiple columns. In the RHS, to compute the `max` on columns specified in `.SDcols`, we make use of the base function `lapply()` along with `.SD` in the same way as we have seen before in the *"Introduction to data.table"* vignette. It returns a list of two elements, containing the maximum value corresponding to `dep_delay` and `arr_delay` for each group.
 
 #
-Before moving on to the next section, let's clean up the newly created columns `speed`, `max_speed`, `max_dep_delay` and `max_arr_delay`.
+Let's clean up the newly created columns `speed`, `max_speed`, `max_dep_delay` and `max_arr_delay`.
```
Comment (MichaelChirico) [inline vignettes/datatable-reference-semantics.Rmd:253]:
```suggestion
Before moving on to the next section, let's clean up the newly created columns `speed`, `max_speed`, `max_dep_delay` and `max_arr_delay`.
```

File: vignettes/datatable-reference-semantics.Rmd (hunk 9 above)
Comment (MichaelChirico) [reply vignettes/datatable-reference-semantics.Rmd:265]: Got it & makes sense, but let's reverse the order -- show the simpler way first, then the more robust way second.

**Comment by MichaelChirico** [comment]
Friendly ping :) would be great to get this merged!

**Comment by ColeMiller1** [comment]
I will make changes tonight or by the end of tomorrow night. Thanks for going through everything.

Cole

File: R/data.table.R (hunk 15 above)
Comment (ColeMiller1) [reply R/data.table.R:1128]: If we wanted to support namespaced calls like `base::names(.SD)`, it seems like doing it from `%iscall%` might be OK. I've seen you guys already specifically address this issue elsewhere, like `data.table::first()`, but it might be handy.

This is only partially tested - I peeked at what you did to fix data.table::first() and the like.

Regardless, I'm including your suggested changes. Thank you.

``` r
"%iscall2%" = function(e, f) {
   ans = FALSE
   if (is.call(e)) {
     test_exp = e[[1L]]

     test_exp_name = if (is.symbol(test_exp)) test_exp else test_exp[[3L]]
     ans = test_exp_name %chin% f
   }
   return(ans)
 }
```

File: inst/tests/tests.Rraw (hunk 19 above)
Comment (ColeMiller1) [reply inst/tests/tests.Rraw:18323]: Got it - I am starting with .01 and incrementing by .01 per your suggestion. It's possible I started with 9 tests and then fudged the numbers at 2247.9 when adding more. Renumbering as you suggest makes the most sense.

File: R/data.table.R (hunk 15 above)
Comment (ColeMiller1) [reply R/data.table.R:1128]: I retract my suggested change to `iscall`. A simpler way would be to do `is.call(q[[1L]]) && deparse(q[[1L]]) %chin% c('names', 'base::names') && ...`  if we wanted to support namespaced calls. My commit will still just be your suggestion for now.

**Comment by ColeMiller1** [comment]
I will give the vignette's another look tomorrow and give you a final ping when I'm done.

File: NEWS.md [hunk 21]
```diff
@@ -18,6 +18,8 @@
 
 3. Namespace-qualifying `data.table::shift()`, `data.table::first()`, or `data.table::last()` will not deactivate GForce, [#5942](https://github.com/Rdatatable/data.table/issues/5942). Thanks @MichaelChirico for the proposal and fix. Namespace-qualifying other calls like `stats::sum()`, `base::prod()`, etc., continue to work as an escape valve to avoid GForce, e.g. to ensure S3 method dispatch.
 
+4. Using `dt[, names(.SD) := lapply(.SD, fx)]` now works, [#795](https://github.com/Rdatatable/data.table/issues/795) -- one of our [most-requested issues (see #3189)](https://github.com/Rdatatable/data.table/issues/3189). Thanks to @brodieG for the report and @ColeMiller1 for PR.
```
Comment (MichaelChirico) [inline NEWS.md:21]:
```suggestion
4. Using `dt[, names(.SD) := lapply(.SD, fx)]` now works, [#795](https://github.com/Rdatatable/data.table/issues/795) -- one of our [most-requested issues (see #3189)](https://github.com/Rdatatable/data.table/issues/3189). Thanks to @brodieG for the report, 20 or so others for chiming in, and @ColeMiller1 for PR.
```

File: R/data.table.R [hunk 22]
```diff
@@ -1122,8 +1122,14 @@ replace_dot_alias = function(e) {
...
-            # e.g. (MyVar):= or get("MyVar"):=
-            lhs = eval(lhs, parent.frame(), parent.frame())
+            # i.e lhs is names(.SD) || setdiff(names(.SD), cols) || (cols)
+            replace_names_sd = function(e, cols){
+              if (length(e) == 1L) return(e)
+              if (e %iscall% "names" && is.name(e2 <- e[[2L]]) && e2 == ".SD") return(cols)
+              for (i in 2:length(e)) if (!is.null(e[[i]])) e[[i]] = replace_names_sd(e[[i]], cols)
+              e
+            }
+            lhs = eval(replace_names_sd(lhs, sdvars), parent.frame(), parent.frame())
```
Comment (MichaelChirico) [inline R/data.table.R:1132]: A simpler implementation might be the following:

```r
e <- copyenv(parent.frame()) # pseudocode
e$.SD <- setNames(logical(length(sdvars)), sdvars) # or vector("list"), or even vector("raw") to really scrimp on storage
lhs = eval(lhs, e, e)
```

WDYT?

File: vignettes/datatable-reference-semantics.Rmd [hunk 23]
```diff
@@ -258,6 +258,23 @@ flights[, c("speed", "max_speed", "max_dep_delay", "max_arr_delay") := NULL]
 head(flights)
 ```
 
+#### -- How can we update multiple existing columns in place using `.SD`?
+
+```{r}
+flights[, names(.SD) := lapply(.SD, as.factor), .SDcols = is.character]
```
Comment (MichaelChirico) [inline vignettes/datatable-reference-semantics.Rmd:264]: Something I can imagine will happen soon is a user trying:

```r
x[, names(.SD)[1:5] := ...]
```

Does that work already? If so, please add a test. If not, no need to handle it until it's requested later unless you see an easy fix.

File: vignettes/datatable-reference-semantics.Rmd [hunk 24]
```diff
@@ -258,6 +258,23 @@ flights[, c("speed", "max_speed", "max_dep_delay", "max_arr_delay") := NULL]
...
+```
+Let's clean up again and make our newly made factor columns back to character columns. This time we will make use of `.SDcols` accepting a function to return columns. In this case, `is.factor` will return the columns which are factors. For more on the **S**ubset of the **D**ata, there is also an [SD Usage vignette](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-sd-usage.html).
```
Comment (MichaelChirico) [inline vignettes/datatable-reference-semantics.Rmd:266]:
```suggestion
Let's clean up again and convert our newly-made factor columns back into character columns. This time we will make use of `.SDcols` accepting a function to decide which columns to include. In this case, `is.factor()` will return the columns which are factors. For more on the **S**ubset of the **D**ata, there is also an [SD Usage vignette](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-sd-usage.html).
```

File: vignettes/datatable-sd-usage.Rmd [hunk 25]
```diff
@@ -77,7 +77,15 @@ The first way to impact what `.SD` is is to limit the _columns_ contained in `.S
...
 ```
 
-This is just for illustration and was pretty boring. But even this simply usage lends itself to a wide variety of highly beneficial / ubiquitous data manipulation operations:
+This is just for illustration and was pretty boring. In addition to accepting a character vector, `.SDcols` also accepts:
+
+1. any function such as `is.character` to filter _columns_
+2. the function^{*} `patterns()` to filter by _column names_
```
Comment (MichaelChirico) [inline vignettes/datatable-sd-usage.Rmd:83]:
```suggestion
2. the function^{*} `patterns()` to filter _column names_ by regular expression
```

File: vignettes/datatable-sd-usage.Rmd [hunk 26]
```diff
@@ -91,52 +99,40 @@ We notice that the following columns are stored as `character` in the `Teams` da
...
 head(unique(Teams[[fkt[1L]]]))
 ```
 
-Note that we must wrap `fkt` in parentheses `()` to force `data.table` to interpret this as column names, instead of trying to assign a column named `'fkt'`.
+Note: 
 
-Actually, the `.SDcols` argument is quite flexible; above, we supplied a `character` vector of column names. In other situations, it is more convenient to supply an `integer` vector of column _positions_ or a `logical` vector dictating include/exclude for each column. `.SDcols` even accepts regular expression-based pattern matching.
+1. The `:=` is an assignment operator to update the `data.table` in place without making a copy. See [reference semantics](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-reference-semantics.html) for more. 
+2. `names(.SD)` indicates which columns we are updating - in this case we update the entire `.SD`.
+3. `lapply()` loops through each column of the `.SD` and converts the column to a factor.
```
Comment (MichaelChirico) [inline vignettes/datatable-sd-usage.Rmd:117]:
```suggestion
2. The LHS, `names(.SD)`, indicates which columns we are updating - in this case we update the entire `.SD`.
3. The RHS, `lapply()`, loops through each column of the `.SD` and converts the column to a factor.
```

**Comment by MichaelChirico** [comment]
Looks great! Last thing is checking if we can simplify implementation / possibly make it much more general.

**Comment by ColeMiller1** [comment]
I attempted to make it more general per your suggestion. I copied the implementation below. All the tests pass and it supports something like `dt[, names(.SD)[1:5] :=...]`. After the commit, I checked to see if `base::names(.SD)` works and it would as well. 

It's easy to revert back or to go down a route you suggested as well. Let me know and I add a `base::names(.SD)` test, too, depending on how we go.

```r
            # i.e lhs is names(.SD) || setdiff(names(.SD), cols) || (cols)
            lhs = eval(lhs, list(.SD = setNames(logical(length(sdvars)), sdvars)), parent.frame())
```

**Comment by TysonStanley** [comment]
I'm a personal fan of the ability to use `dt[, names(.SD)[1:5] :=...]`. Would be a really nice feature to have that flexibility.

**Comment by MichaelChirico** [comment]
New version looks great, awesome!

File: R/data.table.R [hunk 27]
```diff
@@ -1122,8 +1122,8 @@ replace_dot_alias = function(e) {
           if (is.name(lhs)) {
             lhs = as.character(lhs)
           } else {
-            # e.g. (MyVar):= or get("MyVar"):=
-            lhs = eval(lhs, parent.frame(), parent.frame())
+            # i.e lhs is names(.SD) || setdiff(names(.SD), cols) || (cols)
```
Comment (MichaelChirico) [inline R/data.table.R:1125]:
```suggestion
            # lhs is e.g. (MyVar) or get("MyVar") or names(.SD) || setdiff(names(.SD), cols)
```

**Comment by MichaelChirico** [comment]
Almost 10 years later, it's a one-line change :sunglasses:

Great stuff!

**Review by MichaelChirico — APPROVED** [review]
